### DISCUSSION TITLE
[01/20]  Jeonglee 스터디 및 회의예정 내용

### DISCUSSION THREAD

**Discussion by fkdl0048** [BRIDGE-DEV/Unity_Custom_Editor_Study#3] [Ideas]
## 예상 필수 기능  

1. 툴바에서 에디터를 생성하여 panle로 다룰 수 있어야 한다.
2. 에디터 녹화버튼을 통해서 녹화후 시작을 하면 User의 Input(키보드, 마우스)의 입력이 파일형태로 저장된다.
3. 해당되는 값들은 종료 후 수정 또는 삭제가 가능하게 한다.
4. 녹화된 값을 실행을 누르고 실행할 경우 해당 값에 맞게 동작한다.

### 1번
1번은 UI ToolKit을 최대한 활용한다.  
`using UnityEngine.UIElements`라이브러리를 활용(EditorWindow, CreateGUI 등등)  

### 2번  
2번은 user의 Input을 저장할 때 아직 양식은 감이 오지 않지만 그냥 Queue형태로 담고있으면 될 것 같다.  
FIFO형태의 구조여야 입력값이 순서 및 시간대로 정렬이 된다.  

or

Dictionary형태로 key값은 keycode, value은 time으로 저장하여 순서대로 접근하는 방법  
시간, 입력키(이후 추가..? 멀티키?)등을 담고 있는 구조체 or 클래스로 정의하여 담는 방법  

#### 추가적으로 필요해 보이는 기능  

처음에는 모든 키 입력을 다 받게 하고 싶었는데 오히려 역효과가 날 것 같은 생각이 든다.  

과도한 예외처리와 같은 범위를 크게 잡은 느낌..  

에디터딴에서 검증할 키만 세팅하는 기능이 추가되어야 할 것 같음  

### 3번  

에디터에서 입력한 순서대로 시간과 키가 표시되고 수정될 수 있게 만들어야 함  

## 예상 구현 방식

녹화시 씬레벨에 배치가 되어야 게임중의 입력 값을 받을 수 있어서 프리팹을 동적으로 생성하던지 녹화버튼을 통해 생성 삭제를 관리하는 방법.  

총 3가지 덩어리로 본다면 에디터를 관리하는 친구하나, 입력을 감지하고 파싱하는 친구 하나, 입력된 값을 실행하는 친구 하나  

**에디터에 대한 생각 정리**
---  

에디터의 경우 예제가 다양하고 UI디자인 및 연결부분이라 이야기가 필요  

![image](https://user-images.githubusercontent.com/84510455/213551535-ee3aeda8-6382-433b-9fe2-bd13dafabfa4.png)

UI Debugger가 매우매우 잘되어 있음 이벤트 및 실시간 감시 가능  
*eventDebugger마찬가지*  

UI ToolKit디자인 및 배치에 어려움은 없을 것 같음  

Unity프로그램 자체에 Sample이 내장되어 있음  

Window -> UI Toolkit -> Samples  

**키 입력에 대한 생각 정리**
---

1. UI Toolkit `RegisterCallback`사용하기  

매우매우 유용한 기능인 이벤트 콜백이다.  

```cs
using UnityEngine;
using UnityEngine.UIElements;

// Add KeyboardEventTest to a GameObject with a valid UIDocument.
// When the user presses a key, it will print the keyboard event properties to the console.
[RequireComponent(typeof(UIDocument))]
public class KeyboardEventTest : MonoBehaviour
{
    void OnEnable()
    {
        var root = GetComponent<UIDocument>().rootVisualElement;
        root.Add(new Label("Press any key to see the keyDown properties"));
        root.Add(new TextField());
        root.Q<TextField>().Focus();
        root.RegisterCallback<KeyDownEvent>(OnKeyDown, TrickleDown.TrickleDown);
        root.RegisterCallback<KeyUpEvent>(OnKeyUp, TrickleDown.TrickleDown);
    }
    void OnKeyDown(KeyDownEvent ev)
    {
        Debug.Log("KeyDown:" + ev.keyCode);
        Debug.Log("KeyDown:" + ev.character);
        Debug.Log("KeyDown:" + ev.modifiers);
    }

    void OnKeyUp(KeyUpEvent ev)
    {
        Debug.Log("KeyUp:" + ev.keyCode);
        Debug.Log("KeyUp:" + ev.character);
        Debug.Log("KeyUp:" + ev.modifiers);
    }
}
```

실제로 키보드 모든 키가 매핑되는 것을 알 수 있다.  

추가로 수정키기능을 지원하기 때문에 위에서 말한 멀티키 기능도 매우 유용할 듯 하다.  

> modifiers: modifiers 프로퍼티는 누르고 있는 수정 키를 반환  

자세한 이벤트 동작방식은 아래 링크 참고  

++ 이걸로만들어서 동작하게 하고 싶었지만 아쉽게도 아직은 TextField만 작동하는 것 같다.  

분명하게 해결책이 있을텐데 못찾는 중..

[관련 링크](https://docs.unity3d.com/kr/2021.1/Manual/UIE-Events-Handling.html)  

2. if (Input.anyKeyDown)  

위 UI Toolkit방법이 해결이 되지 않는다면 지금까지는 가장 유력한 방법  

Update로 키입력을 실시간으로 감지하고 입력이 되었다면 위에서 말한 에디터에서 감시할 키를 체크 후 일치하다면 해당 키가 눌린 시간과 키를 기록  

즉, Editor에선 사용할 키에 대한 리스트필요 -> 내부에선 해당 리스트에 포함된 키가 읽히면 해당 키를 큐에 저장 후 에디터에 출력  

**동작확인을 위한 간단한 테스트 코드**
```cs
using System.Collections.Generic;
using UnityEngine;

public class KeyInfo
{
    public KeyCode key;
    public float time;

    public KeyInfo(KeyCode key, float time)
    {
        this.key = key;
        this.time = time;
    }
}

public class Test : MonoBehaviour
{
    private List<KeyCode> keycodes;
    Queue<KeyInfo> _queue;
    void Start()
    {
        _queue = new Queue<KeyInfo>();
        keycodes = new List<KeyCode>();
        
        keycodes.Add(KeyCode.A); // 에디터에서 사용할 키를 받아옴
        keycodes.Add(KeyCode.Space);
    }
    
    void Update()
    {
        if (Input.anyKeyDown)
        {
            foreach (var key in keycodes)
            {
                // 키 검증 후 저장
                if (Input.GetKeyDown(key))
                {
                    _queue.Enqueue(new KeyInfo(key, Time.time)); 
                }
            }
        }
    }

    private void OnDisable()
    {
        foreach (var dic in _queue)
        {
            Debug.Log($"{dic.key} : {dic.time}");
        }
    }
}
```

![image](https://user-images.githubusercontent.com/84510455/213551930-ac97a306-81c2-4baf-9bf7-3c88e4ad4d6e.png)

문제없이 동작..!

**키 출력에 대한 생각 정리**  
---  

마찬가지로 위의 방식이 정해지고 나서 고민해봐야할 문제이지만.. 다행이도 키보드 이벤트를 보내는 라이브러리가 존재  

[관련 링크](https://coderzero.tistory.com/entry/%EC%9C%A0%EB%8B%88%ED%8B%B0-%EC%8A%A4%ED%81%AC%EB%A6%BD%ED%8A%B8-%EC%86%8C%EC%8A%A4-%ED%82%A4%EB%B3%B4%EB%93%9C-%EC%9D%B4%EB%B2%A4%ED%8A%B8-%EB%B3%B4%EB%82%B4%EA%B8%B0)  

## 도움이 되는 예제들  

[UI ToolKit Document](https://docs.unity3d.com/Manual/UIE-simple-ui-toolkit-workflow.html)  

[커스텀 패키지 릴리스 튜토리얼](https://www.youtube.com/watch?v=mgsLb3TKljk&ab_channel=Unity)  

[커스텀 패키지 오픈소스 사이트](https://openupm.com/)

## 논의 내용  

1. 멀티키 입력  
2. 꾸욱 누르는 키입력  
3. 사용할 키 입력 방식

**Comment by kugorang** [comment]
정안님이 내용을 잘 적어주셔서 제가 추가로 Dicussion을 생성하지는 않았습니다.

어제 제가 먼저 올려드리기로 했는데, 이번 주는 UI Toolkit에 대한 공부 위주로 진행하여 
다른 오픈소스 분석에 시간 투자를 많이 못 했습니다. 늦어져서 죄송하다는 말씀을 드립니다.

위에 내용을 보고 오늘 논의하면 좋을 내용들을 아래에 몇 개 적습니다.

1. 이 기능을 많이 사용되게 하려면 런타임에서 디버깅을 위한 기능으로 사용되어야 의미있을 것 같습니다.
특히, Unity를 사용하는 대다수의 개발사가 모바일 환경의 프로젝트를 진행하는 것을 감안하면,
UnityEditor 환경 뿐만 아니라 UnityEngine 환경에서도 잘 동작하면 좋겠습니다.

UI Toolkit이 이를 모두 지원하기에 추가적인 작업량은 적을 거라 생각됩니다.

2. 키 입력 항목의 1번 내용의 경우, UI에 대한 키 입력에 한정된 Callback 방식으로 이해했습니다.
그래서 인게임 상 키 입력 감지를 위해서는 2번 방식이 맞다고 보이고,
이를 위해 UniRx 같은 상태 감지 라이브러리를 도입하면 어떨까 하는 의견이 있습니다.

다만, 이를 위해 추가적인 러닝 커브가 필요할 것으로 보여 일정에 딜레이가 생길 경우 적용하지 않아도 됩니다.

3. UnityEngine의 Input System 중, Touch에는 TouchPhase 라는 개념이 존재합니다.

https://docs.unity3d.com/Packages/com.unity.inputsystem@1.4/api/UnityEngine.InputSystem.TouchPhase.html
(New InputSystem의 TouchPase Enum에 대한 내용입니다.)

키보드도 이와 비슷하게 누르기 시작한 시점, 누르고 있는 시점, 누르기 종료된 시점을 구분할 수 있는데,
입력 녹화를 진행할 때 이 Phase까지 같이 기록하게 되면 논의 내용 중,
멀티 키 입력과 Long Pressed 상태까지 모두 처리 가능할 것으로 예상됩니다.
